aea/manager.py

```diff
@@ -0,0 +1,534 @@
+        AsyncTask.__init__(self, agent, loop)
+        self._thread: Optional[Thread] = None
+
+    def create_run_loop(self) -> None:
+        """Create run loop."""
+        self.run_loop = asyncio.new_event_loop()
+
+    def start(self) -> None:
+        """Run task in a dedicated thread."""
+        super().start()
+        self._thread = threading.Thread(
+            target=self.run_loop.run_until_complete, args=[self.task], daemon=True
+        )
+        self._thread.start()
+
+
+class Manager:
+    """Abstract agents manager."""
+
+    AGENT_DO_NOT_OVERRIDE_VALUES = ["skills", "connections", "protocols", "contracts"]
+    MODES = ["async", "threaded"]
+    DEFALUT_TIMEOUT_FOR_BLOCKING_OPERATIONS = 60
+
+    def __init__(self, working_dir: str, mode: str = "async") -> None:
+        """
+        Initialize manager.
+
+        :param working_dir: directory to store base agents.
+        """
+        self.working_dir = working_dir
+        self._was_working_dir_created = False
+        self._is_running = False
+        self._projects: Dict[PublicId, Project] = {}
+        self._keys_dir = os.path.abspath(os.path.join(self.working_dir, "keys"))
+        self._agents: Dict[str, AgentAlias] = {}
+        self._agents_tasks: Dict[str, AsyncTask] = {}
+
+        self._thread = Thread(target=self._run_thread, daemon=True)
+        self._loop: Optional[asyncio.AbstractEventLoop] = None
+        self._event: Optional[asyncio.Event] = None
+
+        self._error_callbacks: List[Callable[[str, BaseException], None]] = []
+
+        if mode not in self.MODES:
+            raise ValueError(
+                f'Invalid mode {mode}. Valid modes are {", ".join(self.MODES)}'
+            )
+        self._started_event = threading.Event()
+        self._mode = mode
+
+    @property
+    def is_running(self) -> bool:
+        """Is manager running."""
+        return self._is_running
+
+    def _run_thread(self) -> None:
+        """Run internal thread with own event loop."""
+        self._loop = asyncio.new_event_loop()
+        self._event = asyncio.Event(loop=self._loop)
+        self._loop.run_until_complete(self._manager_loop())
+
+    async def _manager_loop(self) -> None:
+        """Perform manager stop."""
+        if not self._event:
+            raise ValueError("Do not use this method directly, use start_manager.")
+
+        self._started_event.set()
+
+        while self._is_running:
+            tasks_for_agents = {
+                task.wait(): agent_name
+                for agent_name, task in self._agents_tasks.items()
+            }
+            wait_tasks = list(tasks_for_agents.keys()) + [self._event.wait()]  # type: ignore
+            done, _ = await asyncio.wait(wait_tasks, return_when=FIRST_COMPLETED)
+
+            if self._event.is_set():
+                self._event.clear()
+
+            for task in done:
+                if task not in tasks_for_agents:
+                    await task
```

Comment:
here we don't handle exceptions (due to `continue`), is that intended?

Reply:
What kind of tasks do we expect here; please add comment

Reply:
done.
exception handled as expected, in the next lines